sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -653,6 +655,128 @@ case class StringRPad(str: Expression, len: Expression, pad: Expression)
 }
 
 /**
+ * Extracts a part from a URL
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(url, partToExtract[, key]) - extracts a part from a URL",
+  extended = "Parts: HOST, PATH, QUERY, REF, PROTOCOL, AUTHORITY, FILE, USERINFO\n"
+  + "key specifies which query to extract\n"
+  + "Examples:\n"
+  + "  > SELECT _FUNC_('http://spark.apache.org/path?query=1', "
+  + "'HOST') FROM src LIMIT 1;\n" + "  'spark.apache.org'\n"
+  + "  > SELECT _FUNC_('http://spark.apache.org/path?query=1', "
+  + "'QUERY') FROM src LIMIT 1;\n"  + "  'query=1'\n"
+  + "  > SELECT _FUNC_('http://spark.apache.org/path?query=1', "
+  + "'QUERY', 'query') FROM src LIMIT 1;\n" + "  '1'")
+case class ParseUrl(children: Expression*)
+  extends Expression with ImplicitCastInputTypes with CodegenFallback {
+
+  override def nullable: Boolean = true
+
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+
+  private lazy val stringExprs = children.toArray
+  @transient private var lastUrlStr: UTF8String = _
+  @transient private var lastUrl: URL = _
+  // last key in string, we will update the pattern if key value changed.
+  @transient private var lastKey: UTF8String = _
+  // last regex pattern, we cache it for performance concern
+  @transient private var pattern: Pattern = _
+  private lazy val HOST: UTF8String = UTF8String.fromString("HOST")
+  private lazy val PATH: UTF8String = UTF8String.fromString("PATH")
+  private lazy val QUERY: UTF8String = UTF8String.fromString("QUERY")
+  private lazy val REF: UTF8String = UTF8String.fromString("REF")
+  private lazy val PROTOCOL: UTF8String = UTF8String.fromString("PROTOCOL")
+  private lazy val FILE: UTF8String = UTF8String.fromString("FILE")
+  private lazy val AUTHORITY: UTF8String = UTF8String.fromString("AUTHORITY")
+  private lazy val USERINFO: UTF8String = UTF8String.fromString("USERINFO")
+  private lazy val REGEXPREFIX: String = "[&^]?"
+  private lazy val REGEXSUBFIX: String = "=([^&]*)"
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      TypeCheckResult.TypeCheckFailure("parse_url function requires two or three arguments")
+    } else {
+      super[ImplicitCastInputTypes].checkInputDataTypes()
+    }
+  }
+
+  def parseUrlWithoutKey(url: Any, partToExtract: Any): Any = {
+    if (url == null || partToExtract == null) {
+      null
+    } else {
+      if (lastUrlStr == null || !url.equals(lastUrlStr)) {
+        try {
+          lastUrlStr = url.asInstanceOf[UTF8String].clone()
+          lastUrl = new URL(url.toString)
+        } catch {
+          case ex: MalformedURLException => return null
```

Comment:
We can use the following like Hive uses `Exception`. `SecurityException` occurs possibly as a `RuntimeException`.
```
case NonFatal(_) =>
```